proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,946 @@
+  versions of the resource, each with a different name, even though the
+  contents are the same, which can also cause unnecessarily poor cache
+  performance.  For example, in the "dynamic route selection" use-case,
+  let's say that every client uses two different dynamic selection
+  parameters, `env` (which can have one of the values `prod`, `canary`, or
+  `test`) and `version` (which can have one of the values `v1`, `v2`, or
+  `v3`).  Now let's say that there is a `RouteConfiguration` with one route
+  that should be selected via the parameter `env=prod` and another route that
+  should be selected via the parameter `version=v1`. This means that there
+  are only four variants of the `RouteConfiguration` resource (`{env!=prod,
+  version!=v1}`, `{env=prod, version!=v1}`, `{env!=prod, version=v1}`, and
+  `{env=prod, version=v1}`).  However, the exact-match semantics means
+  that there will have to be nine different versions of this resource,
+  one for each combination of values of the two parameters.
+
+### Related Proposals:
+* [xRFC TP1: new xDS naming scheme](TP1-xds-transport-next.md)
+
+## Proposal
+
+This document proposes an alternative approach.  We start with the
+observation that resource names are used in two places:
+
+- The **transport protocol** layer, which needs to identify the right
+  resource contents to send for a given resource name, often obtaining
+  those resource contents from a cache.
+- The **resource graph** used on an individual client, where there are a
+  set of data model resources that refer to each other by name.  For
+  example, a `RouteConfiguration` refers to individual `Cluster` resources
+  by name.
+
+The use-cases that we're aware of for dynamic resource selection have
+an important property that we can take advantage of.  When multiple
+variants of a given resource exist, any given client will only ever use
+one of those variants at a given time.  That means that the parameters
+that affect which variant of the resource is used are required by the
+transport protocol, but they are not required by the client's data model.
+
+It should be noted that caching xDS proxies, unlike "leaf" clients, will
+need to track multiple variants of each resource, since a given caching
+proxy may be serving clients that need different variants of a given
+resource.  However, since caching xDS proxies deal with resources only
+at the transport protocol layer, the resource graph layer is
+essentially irrelevant in that case.
+
+### Dynamic Parameters
+
+With the above property in mind, this document proposes the following
+data structures:
+- **Dynamic parameters**, which are a set of key/value pairs sent by the
+  client when subscribing to a resource.
+- **Dynamic parameter constraints**, which are a set of criteria that
+  can be used to determine whether a set of dynamic parameters matches
+  the constraints.  These constraints are part of the cache key for an
```

Comment:
Is it correct to say that this is part of the cache key or rather that the constraints are used during the cache lookup process? The distinction is that a key is usually a pure transformation of some input to a key space, here we are doing something more like constraint satisfaction.

Reply:
Reading further, I think the ambiguity is on what the cache key is. Is this a cache keyed by the resource names / constraints provided by the server, or a cache for the client, keyed by the dynamic parameters provided by the client? It might be worth clarifying this in various places to reduce reader potential for confusion.

Reply:
The intent here is that the cache key is both the resource name and the constraints.  It has to be that way, because we need to cache each variant of the resource separately, and each variant is uniquely identified by its name and constraints.  See the "Considerations for Implementations" section below for more details.

Do you have any specific suggestions as to changes I could make to clarify this?

Reply:
Can you add an explicit definition of cache key somewhere?

Reply:
I've changed the wording to avoid the term "cache key".  Does that address your concern?

Reply:
Yep, thanks.